streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -971,6 +971,130 @@ trait Stream_Functions {
   final def dieMessage(msg: String): ZStream[Any, Nothing, Nothing] =
     halt(Cause.die(new RuntimeException(msg)))
 
+  /**
+   * Creates a stream from an asynchronous callback that can be called multiple times.
+   */
+  final def effectAsync[R: ConformsR, E, A](
+    register: (ZIO[R, E, A] => Unit) => Unit,
+    outputBuffer: Int = 64
+  ): ZStream[R, E, A] =
+    new ZStream[R, E, A] {
+      override def fold[R1 <: R, E1 >: E, A1 >: A, S]: Fold[R1, E1, A1, S] =
+        ZManaged.succeedLazy { (s, cont, g) =>
+          for {
+            output  <- Queue.bounded[Take[E1, A1]](outputBuffer).toManaged(_.shutdown)
+            runtime <- ZIO.runtime[R].toManaged_
+            _ <- ZManaged.succeedLazy {
+                  register(
+                    k =>
+                      runtime.unsafeRunAsync_(
+                        k.foldCauseM(
+                          cause => output.offer(Take.Fail(cause)).unit,
+                          b => output.offer(Take.Value(b)).unit
+                        )
+                      )
+                  )
+                }
+            s <- ZStream.fromQueue(output).unTake.fold[R1, E1, A1, S].flatMap(fold => fold(s, cont, g))
+          } yield s
+        }
+    }
+
+  /**
+   * Creates a stream from an asynchronous callback that can be called multiple times.
+   * The registration of the callback can possibly return the stream synchronously
+   */
+  final def effectAsyncMaybe[R: ConformsR, E, A](
+    register: (ZIO[R, E, A] => Unit) => Option[ZStream[R, E, A]],
+    outputBuffer: Int = 64
+  ): ZStream[R, E, A] =
+    new ZStream[R, E, A] {
+      override def fold[R1 <: R, E1 >: E, A1 >: A, S]: Fold[R1, E1, A1, S] =
+        ZManaged.succeedLazy { (s, cont, g) =>
+          for {
+            output  <- Queue.bounded[Take[E1, A1]](outputBuffer).toManaged(_.shutdown)
+            runtime <- ZIO.runtime[R].toManaged_
+            maybeStream <- ZManaged.succeedLazy {
+                            register(
+                              k =>
+                                runtime.unsafeRunAsync_(
+                                  k.foldCauseM(
+                                    cause => output.offer(Take.Fail(cause)).unit,
+                                    b => output.offer(Take.Value(b)).unit
+                                  )
+                                )
+                            )
+                          }
+            s <- maybeStream match {
+                  case Some(stream) => stream.fold[R1, E1, A1, S].flatMap(fold => fold(s, cont, g))
```

Comment:
For safety, I think we should eagerly shutdown the output queue in this case. Should surface implementation errors sooner.

Reply:
I just noticed you asked in the PR description if you should only construct the queue when needed; that'd be optimal indeed but I don't immediately see how to do that.